golang/cosmos/ante/inbound.go

```diff
@@ -0,0 +1,73 @@
+package ante
+
+import (
+	"github.com/armon/go-metrics"
+
+	"github.com/cosmos/cosmos-sdk/telemetry"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	swingtypes "github.com/Agoric/agoric-sdk/golang/cosmos/x/swingset/types"
+)
+
+// TODO: We don't have a more appropriate error type for this.
+var ErrInboundQueueFull = sdkerrors.ErrMempoolIsFull
+
+type inboundAnte struct {
+	sk SwingsetKeeper
+}
+
+// NewInboundDecorator return an AnteDecorator which honors the allowed size of the inbound queue.
+// The swingset message types will consume one allowed entry each, and will be reflected
+// in the action queue when executed.
+func NewInboundDecorator(sk SwingsetKeeper) sdk.AnteDecorator {
+	return inboundAnte{sk: sk}
+}
+
+// AnteHandle implements sdk.AnteDecorator
+func (ia inboundAnte) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
+	msgs := tx.GetMsgs()
+	inboundsAllowed := int32(-1)
+	for _, msg := range msgs {
+		switch msg.(type) {
+		case *swingtypes.MsgDeliverInbound,
+			*swingtypes.MsgInstallBundle,
+			*swingtypes.MsgProvision,
+			*swingtypes.MsgWalletAction,
+			*swingtypes.MsgWalletSpendAction:
+			if inboundsAllowed == -1 {
+				state := ia.sk.GetState(ctx)
+				allowed, found := swingtypes.QueueSizeEntry(state.QueueAllowed, "inbound")
+				if found {
+					actions, err := ia.sk.ActionQueueLength(ctx)
+					if err != nil {
+						return ctx, err
+					}
+					if actions < allowed {
+						inboundsAllowed = allowed - actions
+					} else {
+						inboundsAllowed = 0
+					}
+				} else {
+					// no inbound allowed size from swingset
+					inboundsAllowed = 0
+				}
+			}
+			if inboundsAllowed > 0 {
+				inboundsAllowed--
```

Comment:
This looks like it's local to this function/method, and it looks like it's called with a single txn (which might contain multiple messages). Do ante handlers get run for all txns before any of them are processed? Or does it do `ante handler, process txn` as a pair, before moving on to the next one? If it's the former, then a local (ephemeral) `inboundsAllowed` would reset for each txn, and allow in too many.

Reply:
I'm taking a look, but in the worst case, we let too many txns in, and then in the *next* block we reject them all, giving us a little hysteresis, which isn't necessarily a bad thing.

Reply:
Confirmed that the ante handler runs at `DeliverTx` time, so we're seeing the state mutated by previous Txs in the block. So we're going to reject the Tx if it has enough swingset Msgs to overflow the inbound queue (counting the pending action queue).